src/services/playlist.ts

```diff
@@ -4,15 +4,265 @@
+                    const subscription = this.m_subscriptions.get (channelId);
+                    if (subscription) {
+                        subscription.unsubscribe ();
+                        this.m_subscriptions.delete (channelId);
+                    }
+                } else {
+                    if (!this.m_subscriptions.has (channelId)) {
+                        const subscription = connection.subscribe (this.m_audioPlayer);
+                        if (subscription) {
+                            this.m_subscriptions.set (channelId, subscription);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    public connect (channel : VoiceChannel) : void {
+        const connection = joinVoiceChannel({
+            channelId: channel.id,
+            guildId: channel.guild.id,
+            adapterCreator: channel.guild.voiceAdapterCreator
+        })
+
+        connection.on (VoiceConnectionStatus.Disconnected, () => {
+            setTimeout(() => {
+                this.connect(channel);
+            }, config.player.rejoinDelay)
+        });
+
+        connection.on ('stateChange', (previousState, currentState) => {
+            // this.info (`connection to ${channel.name} (${channel.guild.name}) `
+            // + `changed from ${previousState.status} to ${currentState.status}`);
+        });
+
+        if (this.getListenerCount (channel) > 0) {
+            connection.subscribe (this.m_audioPlayer);
+        }
+    }
+
+    private async fetchAllChannels () : Promise<VoiceChannel[]> {
+        const guilds = await this.m_databaseService.client.guild.findMany();
+        const res : VoiceChannel[] = [];
+        for (const dbGuild of guilds) {
+            try {
+                const guild = await this.m_discordService.client.guilds.fetch (dbGuild.guildId);
+                const channel = await guild.channels.fetch (dbGuild.channelId);
+                if (channel instanceof VoiceChannel) {
+                    res.push (channel);
+                }
+            } catch (error) {
+                this.error (`Unable to fetch a guild channel: ${error}`);
+            }
+        }
+        return res;
+    }
+
+    private async connectAll () : Promise<void> {
+        const channels = await this.fetchAllChannels();
+        for (const channel of channels) {
+            this.connect (channel);
+        }
+    }
+
+    public getCurrentSong () : Song | null {
+        return this.m_nowPlaying;
+    }
+
+    /**
+     * @returns `true` if the song has been skipped successfully
+     */
+    public skipCurrentSong () : boolean {
+        const song = this.getCurrentSong ();
+        if (song === null) {
+            return false;
+        } else {
+            this.m_announcement = true;
+            this.m_skipVotes.clear ();
+            return this.m_audioPlayer.stop ();
+        }
+    }
+
+    /**
+     * @returns `true` if the number of votes has reached the skip threshold
```

Comment:
no, it returns something else

Reply:
in fact the number of required votes is actually checked in a completely different place, maybe should be checked here instead?